secrets/src/int/secret_integers.rs

```diff
@@ -0,0 +1,235 @@
+use super::classify::*;
+use crate::traits::*;
+use std::ops::*;
+
+pub type I8 = Secret<i8>;
+pub type U8 = Secret<u8>;
+pub type I16 = Secret<i16>;
+pub type U16 = Secret<u16>;
+pub type I32 = Secret<i32>;
+pub type U32 = Secret<u32>;
+pub type I64 = Secret<i64>;
+pub type U64 = Secret<u64>;
+pub type I128 = Secret<i128>;
+pub type U128 = Secret<u128>;
+
+// Construct a secret integer
+pub const fn secret<T>(x: T) -> Secret<T> {
+    Secret(x)
+}
+
+impl<T: Scalar> Classify for T {
+    type Classified = Secret<T>;
+    fn classify(self) -> Secret<Self> {
+        secret(self)
+    }
+}
+
+impl<T: Scalar> Declassify for Secret<T> {
+    type Declassified = T;
+    fn declassify(self) -> T {
+        self.0
+    }
+}
+
+impl<'a, T: Scalar> ClassifyRef for &'a T {
+    type ClassifiedRef = &'a Secret<T>;
+    fn classify_ref(self) -> &'a Secret<T> {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+
+impl<'a, T: Scalar> DeclassifyRef for &'a Secret<T> {
+    type DeclassifiedRef = &'a T;
+    fn declassify_ref(self) -> &'a T {
+        unsafe { core::mem::transmute(self) }
+    }
+}
+
+pub fn classify_mut_slice<T: Scalar>(x: &mut [T]) -> &mut [Secret<T>] {
+    unsafe { core::mem::transmute(x) }
+}
+
+impl<T: Add, V: Into<Secret<T>>> Add<V> for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn add(self, rhs: V) -> Self::Output {
+        self.0.add(rhs.into().0).into()
+    }
+}
+
+impl<T: Sub, V: Into<Secret<T>>> Sub<V> for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn sub(self, rhs: V) -> Self::Output {
+        self.0.sub(rhs.into().0).into()
+    }
+}
+
+impl<T: Mul, V: Into<Secret<T>>> Mul<V> for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn mul(self, rhs: V) -> Self::Output {
+        self.0.mul(rhs.into().0).into()
+    }
+}
+
+impl<T: BitXor, V: Into<Secret<T>>> BitXor<V> for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn bitxor(self, rhs: V) -> Self::Output {
+        self.0.bitxor(rhs.into().0).into()
+    }
+}
+
+impl<T: BitOr, V: Into<Secret<T>>> BitOr<V> for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn bitor(self, rhs: V) -> Self::Output {
+        self.0.bitor(rhs.into().0).into()
+    }
+}
+
+impl<T: BitAnd, V: Into<Secret<T>>> BitAnd<V> for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn bitand(self, rhs: V) -> Self::Output {
+        self.0.bitand(rhs.into().0).into()
+    }
+}
+
+impl<T: Not> Not for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn not(self) -> Self::Output {
+        self.0.not().into()
+    }
+}
+
+impl<U, T: Shl<U>> Shl<U> for Secret<T>
+where
+    T::Output: Into<T>,
+{
+    type Output = Secret<T>;
+    fn shl(self, rhs: U) -> Self::Output {
+        (self.0.shl(rhs).into()).into()
+    }
+}
```

Comment:
More out of curiosity than because I think that would be needed for this to merge, but would we also want to have an impl that is something like this:

```rust
impl<U, T: Shl<U>> Shl<Secret<U>> for Secret<T>
where
    T::Output: Into<T>,
```
i.e. where the shift count operant also is secret?

Reply:
The value we shift things by is explicitly not secret, because we cannot guarantee constant time behavior for it.